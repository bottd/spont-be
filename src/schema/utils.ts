import * as knex from 'knex';
const axios = require('axios');
const environment = process.env.NODE_ENV || 'development';
const config = require('../../knexfile')[environment];
const database = knex(config);

interface Location {
  location_name: string;
  category: string;
  latitude: number;
  longitude: number;
}

interface User {
  id: string;
}

export function selectAllLocations() {
  return database('locations').select();
}

export function selectUserByID(id: string) {
  return database('users')
    .where('id', id)
    .first();
}

export function selectLocationByID(id: string) {
  return database('locations')
    .where('id', id)
    .first();
}

export async function selectLocationsByUserID(id: string) {
  const locationJoins = await database('user_locations')
    .where('user_id', id)
    .select();
  const locations = locationJoins.map(join =>
    selectLocationByID(join.location_id),
  );
  return Promise.all(locations);
}

export async function selectUsersByLocationID(id: string) {
  const userJoins = await database('user_locations')
    .where('location_id', id)
    .select();
  const users = userJoins.map(join => selectUserByID(join.user_id));
  return Promise.all(users);
}
export async function createNewUser() {
  const newUser = await database('users').insert({}, '*');
  return newUser[0];
}

export async function insertLocation(location: Location, user: User) {
  let locationID;
  const checkLocation = await database('locations')
    .where('latitude', location.latitude)
    .andWhere('longitude', location.longitude)
    .first();
  if (!checkLocation) {
    locationID = await database('locations').insert(location, 'id');
  } else {
    locationID = checkLocation.id;
  }
  const checkJoin = await database('user_locations')
    .where('location_id', locationID)
    .andWhere('user_id', user.id)
    .first();
  if (checkJoin) {
    await database('user_locations')
      .where('location_id', locationID)
      .andWhere('user_id', user.id)
      .increment('visit_count', 1);
  } else {
    await database('user_locations').insert(
      {
        location_id: locationID,
        user_id: user.id,
        visit_count: 1,
      },
      '*',
    );
  }
  const returnStuff = {
    userID: user.id,
    latitude: location.latitude,
    longitude: location.longitude,
  };
  return returnStuff;
}

export async function getLocationByCoords(latitude: number, longitude: number) {
  const response = await axios.get(
    `https://maps.googleapis.com/maps/api/place/nearbysearch/json?key=${
      process.env.API_KEY
    }&location=${latitude},${longitude}&rankby=distance`,
  );
  return response.data.results.reduce((array, result) => {
    if (!result.types.includes('route')) {
      array.push({
        category: result.types[0],
        location_name: result.name,
        latitude: result.geometry.location.lat,
        longitude: result.geometry.location.lng,
      });
    }
    return array;
  }, []);
}
