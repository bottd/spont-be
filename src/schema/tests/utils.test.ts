import * as knex from 'knex';
import * as utils from '../utils';
const environment = process.env.NODE_ENV || 'development';
const config = require('../../../knexfile')[environment];
const database = knex(config);
interface Location {
  location_name: string;
  category: string;
  latitude: number;
  longitude: number;
}

interface User {
  id: string;
}

describe('utils methods', () => {
  it('Should match snapshot', () => {
    expect(utils).toMatchSnapshot();
  });

  describe('selectAllLocations', () => {
    beforeAll(async done => {
      await database.migrate.rollback();
      await database.migrate.latest();
      await database.seed.run();
      done();
    });
    it('should return all locations in locations table', async done => {
      const result = await utils.selectAllLocations();
      expect(result.length).toBe(6);
      done();
    });
  });

  describe('selectUserByID', () => {
    beforeAll(async done => {
      await database.migrate.rollback();
      await database.migrate.latest();
      await database.seed.run();
      done();
    });
    it('Should return a user object matching given ID', async done => {
      const users = await database('users').select();
      const selectedUser = await utils.selectUserByID(users[0].id);
      expect(selectedUser).toEqual(users[0]);
      done();
    });
  });

  describe('selectLocationByID', () => {
    beforeAll(async done => {
      await database.migrate.rollback();
      await database.migrate.latest();
      await database.seed.run();
      done();
    });
    it('Should return a location object matching the given ID', async done => {
      const locations = await utils.selectAllLocations();
      const selectedLocation = await utils.selectLocationByID(locations[0].id);
      expect(selectedLocation).toEqual(locations[0]);
      done();
    });
  });

  describe('selectLocationsByUserID', () => {
    beforeAll(async done => {
      await database.migrate.rollback();
      await database.migrate.latest();
      await database.seed.run();
      done();
    });

    it('Should return a set of locations associated with a user id', async done => {
      const users = await database('users').select();
      const userLocations = await utils.selectLocationsByUserID(users[0].id);
      const locations = userLocations.map((location: Location) => ({
        category: location.category,
        location_name: location.location_name,
        latitude: location.latitude,
        longitude: location.longitude,
      }));
      expect(locations.length).toBe(2);
      done();
    });
  });

  describe('selectUsersByLocationID', () => {
    it('Should return a set of locations associated with a user id', async done => {
      const locations = await database('locations').select();
      const users = await utils.selectUsersByLocationID(locations[0].id);
      expect(users.length).toBe(1);
      done();
    });
  });

  describe('createNewUser', () => {
    beforeAll(async done => {
      await database.migrate.rollback();
      await database.migrate.latest();
      await database.seed.run();
      done();
    });
    it('Should return a new UUID with length of 36 and timestamps', async done => {
      const newUser = await utils.createNewUser();
      expect(newUser.id.length).toBe(36);
      expect(newUser.created_at).toBeDefined();
      expect(newUser.updated_at).toBeDefined();
      done();
    });
  });
  describe('insertLocation', () => {
    beforeAll(async done => {
      await database.migrate.rollback();
      await database.migrate.latest();
      await database.seed.run();
      done();
    });
    it('Should insert a user_location join for a user', async done => {
      const users = await database('users').select();
      const allLocations = await utils.selectAllLocations();
      const userLocations = await utils.selectLocationsByUserID(users[1].id);
      await utils.insertLocation(allLocations[0], users[1]);
      const userLocationsNext = await utils.selectLocationsByUserID(users[1].id);
      expect (userLocations.length).not.toBe(userLocationsNext.length);
      done();
    });
  });

  describe('getLocationByCoords', () => {});
  afterAll(async done => {
    await database.migrate.rollback();
    done();
  });
});
