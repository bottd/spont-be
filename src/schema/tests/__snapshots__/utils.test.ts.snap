// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils methods Should match snapshot 1`] = `
Object {
  "createNewUser": [Function],
  "getLocationByCoords": [Function],
  "insertLocation": [Function],
  "selectAllLocations": [Function],
  "selectLocationByID": [Function],
  "selectLocationsByUserID": [Function],
  "selectUserByID": [Function],
  "selectUsersByLocationID": [Function],
}
`;
